sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_asset_utils.py

```diff
@@ -301,96 +295,173 @@ def _get_file_list_content_hash(file_list) -> str:
-    project_dir = root[len(str(working_dir)) :] + "/"
-    file_paths = [
-        convert_windows_path_to_unix(os.path.join(root, name))
-        for name in files
-        if not ignore_file.is_file_excluded(os.path.join(root, name))
-    ]  # get all files not excluded by the ignore file
-    file_paths_including_links = {fp: None for fp in file_paths}
-
-    for path in file_paths:
-        target_prefix = ""
-        symlink_prefix = ""
-
-        # check for symlinks to get their true paths
-        if os.path.islink(path):
-            target_absolute_path = os.path.join(working_dir, os.readlink(path))
-            target_prefix = "/".join([root, str(os.readlink(path))]).replace(project_dir, "/")
-
-            # follow and add child links if the directory is a symlink
-            if os.path.isdir(target_absolute_path):
-                symlink_prefix = path.replace(root + "/", "")
-
-                for r, _, f in os.walk(target_absolute_path, followlinks=True):
-                    target_file_paths = {
-                        os.path.join(r, name): symlink_prefix + os.path.join(r, name).replace(target_prefix, "")
-                        for name in f
-                    }  # for each symlink, store its target_path as key and symlink path as value
-                    file_paths_including_links.update(target_file_paths)  # Add discovered symlinks to file paths list
-            else:
-                file_path_info = {
-                    target_absolute_path: path.replace(root + "/", "")
-                }  # for each symlink, store its target_path as key and symlink path as value
-                file_paths_including_links.update(file_path_info)  # Add discovered symlinks to file paths list
-            del file_paths_including_links[path]  # Remove original symlink entry now that detailed entry has been added
-
-    file_paths = sorted(
-        file_paths_including_links
-    )  # sort files to keep consistent order in case of repeat upload comparisons
-    dir_parts = [convert_windows_path_to_unix(os.path.relpath(root, source)) for _ in file_paths]
-    dir_parts = ["" if dir_part == "." else dir_part + "/" for dir_part in dir_parts]
-    blob_paths = []
-
-    for (dir_part, name) in zip(dir_parts, file_paths):
-        if file_paths_including_links.get(
-            name
-        ):  # for symlinks, use symlink name and structure in directory to create remote upload path
-            blob_path = prefix + dir_part + file_paths_including_links.get(name)
-        else:
-            blob_path = prefix + dir_part + name.replace(root + "/", "")
-        blob_paths.append(blob_path)
+    source_path = Path(source).resolve()
+    prefix = "" if dest == "" else dest + "/"
+    prefix += os.path.basename(source_path) + "/"
+
+    local_upload_paths, link_dict = get_local_paths(source_path=source_path, ignore_file=ignore_file)
+    upload_pairs = construct_remote_paths(local_upload_paths, source_path, prefix, link_dict)
```

Comment:
Thank you for breaking into smaller functions. This made the code much more readable.